### ISSUE TITLE
Setting credentials in 1.1.0-alpha.1 on AWS

### ISSUE THREAD

**Issue by tombenner** [kubernetes/kubernetes#14335]
We're working on upgrading from 1.0 to 1.1.0-alpha.1. Before spinning up a cluster with kube-up, we set the credentials:

```sh
kubectl config set-credentials aws_kubernetes --username=myusername --password=mypassword
```

This works as expected on 1.0. With 1.1.0-alpha.1, though, we saw that these credentials (myusername:mypassword) are overwritten with values created in `gen-kube-basicauth` (admin:randomstring). These credentials from `gen-kube-basicauth` are stored in `.kube/config` and are used in the cluster.

If we comment out `gen-kube-basicauth` in `cluster/aws/util.sh` and set the following env variables, the credentials are set on the cluster as expected:

```sh
export KUBE_USER=myusername
export KUBE_PASSWORD=mypassword
```

Here's our kubectl version:

```sh
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-alpha.1", GitCommit:"9bae6636d5fcf436a3b14d219ced195e10e21fe8", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-alpha.1", GitCommit:"9bae6636d5fcf436a3b14d219ced195e10e21fe8", GitTreeState:"clean"}
```

Is this a bug? Or is there a new strategy for setting credentials in 1.1.0?

**Comment by liggitt** [comment]
the setup scripts probably shouldn't be overwriting existing values in shared config files

**Comment by jlowdermilk** [comment]
We explicitly stopped loading existing credentials in the startup scripts so they are not accidentally reused across providers/clusters. It would be reasonable to re-add the ability to load existing credentials from kubeconfig, so long as we explicitly define a context, not use whatever context happened to be the current one.

**Comment by jlowdermilk** [comment]
I don't think we ever formalized/documented what the expected flow should be for setting/reusing credentials with `kube-up.sh`. Reusing a kubeconfig credential entry in its entirety isn't possible; you can specify user,password (and I think token?) at cluster creation time, but the certs are generated on the master and have to be copied down. To my knowledge, there isn't yet support for providing existing certs at creation time.

We can have [gen-kube-basicauth](https://github.com/kubernetes/kubernetes/blob/master/cluster/common.sh#L163) check for and use existing `KUBE_USER`|`KUBE_PASSWORD`, but I don't think that's a great way to do it. Propose the following approach:

1. In kube-up, check for a defined `KUBE_CONTEXT`.
1. If `KUBE_CONTEXT` is not provided, set it to the default as determined by provider scripts (e.g. [this](https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/util.sh#L1018) for aws).
1. Use `KUBE_CONTEXT` to try to load credentials from kubeconfig. If there is no such context entry in kubeconfig, generate credentials.
1. Spin up cluster resources.
1. Save cluster endpoint info and credentials to kubeconfig under `KUBE_CONTEXT`; this would overwrite the existing entry, possibly with (some of) the same info that was already there.

@kubernetes/goog-control-plane for thoughts.

**Comment by roberthbailey** [comment]
> the certs are generated on the master and have to be copied down.

This is no longer true (at least on GCE). We [generate the certs on the client](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/util.sh#L517-L550) and push them up to the master via the metadata server.

**Comment by jlowdermilk** [comment]
Fixed by #16345, and fix made it into 1.1.1, so closing this. If you want to spin up a cluster that uses existing kubeconfig credentials, export the context to use as `KUBE_CONTEXT` before running `kube-up.sh`

**Comment by tombenner** [comment]
Great, thanks!
